### DISCUSSION TITLE
How does JobRunr store the actual Job Code?

### DISCUSSION THREAD

**Discussion by AleksandarTokarev** [jobrunr/jobrunr#579] [Q&A] [answered]
I am testing Job Runr and i see that Jobs are stored as JSON in `jobrunr_jobs` table. 
Here is the example JSON for the job (Normal Job)
```
{
    "version": 3,
    "jobSignature": "org.jobrunr.examples.services.MyServiceInterface.doLoopJob(java.lang.String)",
    "jobName": "org.jobrunr.examples.services.MyServiceInterface.doLoopJob(Hello World)",
    "jobDetails":
    {
        "className": "org.jobrunr.examples.services.MyServiceInterface",
        "staticFieldName": null,
        "methodName": "doLoopJob",
        "jobParameters": [
        {
            "className": "java.lang.String",
            "actualClassName": "java.lang.String",
            "object": "Hello World"
        }],
        "cacheable": false
    },
    "id": "cbc44a91-4f4e-4818-8235-cd988c4cf599",
    "jobHistory": [
    {
        "@class": "org.jobrunr.jobs.states.EnqueuedState",
        "state": "ENQUEUED",
        "createdAt": "2022-10-03T10:33:52.439339800Z"
    },
    {
        "@class": "org.jobrunr.jobs.states.ProcessingState",
        "state": "PROCESSING",
        "createdAt": "2022-10-03T10:33:55.189854Z",
        "serverId": "520db0b6-024f-4cfa-93d4-d0dd23603ff8",
        "updatedAt": "2022-10-03T10:33:55.189854Z"
    },
    {
        "@class": "org.jobrunr.jobs.states.SucceededState",
        "state": "SUCCEEDED",
        "createdAt": "2022-10-03T10:33:55.196854500Z",
        "latencyDuration": 2.750514200,
        "processDuration": 0.007000500
    }],
    "metadata":
    {
        "@class": "java.util.concurrent.ConcurrentHashMap",
        "jobRunrDashboardLog-2":
        {
            "@class": "org.jobrunr.jobs.context.JobDashboardLogger$JobDashboardLogLines",
            "logLines": []
        }
    }
}
```
My question is the following, what happens if the sending of the jobs is done in one service (lets say 1 repo), and the code for `org.jobrunr.examples.services.MyServiceInterface.doLoopJob` is in that service.
We have other service which is actually processing the Jobs but we dont have the `doLoopJob` code there.
Will that work? Is the actual method code serialized somewhere? Or we have to include that class/method there as well.
Best Regards

**Comment by rdehuyss** [comment, marked as answer]
See the website:

> Important: all your servers must run the same version of your code! If your webapp server has a newer version with a method signature that is not compatible with the server that processes your background jobs, a NoSuchMethod Exception will be thrown and job processing will fail!
As of JobRunr v1.1.0, the dashboard shows an error if there are jobs which cannot be found.


It means that JobRunr does not serialise any bytecode - your code should be available where you enqueue it (or at least an interface) and your code should also be available in the JVM that will do the actual processing.

**Reply by AleksandarTokarev** [reply]
Great, this makes it clear. This means that even the packages have to be the same as the method that will be invoked via Reflection is stored  in the format below. 
```
"jobSignature": "org.jobrunr.examples.services.MyServiceInterface.doLoopJob(java.lang.String)",
"jobName": "org.jobrunr.examples.services.MyServiceInterface.doLoopJob(Hello World)",
```
I assume there is no way for them to be in different packages (via config or something)?

**Reply by rdehuyss** [reply]
No, they indeed must stay in the same package to keep things easy.
